src/apply_fullupdate.jl

```diff
@@ -0,0 +1,231 @@
+
+"""
+Check env is the correct environment for the subset of vertices of tn
+"""
+function assert_correct_environment(ψ::ITensorNetwork, env::Vector{ITensor}, verts::Vector)
+  outer_verts_indices = flatten([commoninds(ψ, e) for e in boundary_edges(ψ, verts)])
+  return issetequal(noncommoninds(env...), outer_verts_indices)
+end
+
+function create_b(
+  p::ITensor,
+  q::ITensor,
+  gate::ITensor,
+  envs::Vector{ITensor},
+  bottom_tensor::ITensor;
+  opt_sequence=nothing,
+)
+  return noprime(
+    ITensors.contract(
+      vcat(ITensor[p, q, gate, dag(prime(bottom_tensor))], envs); sequence=opt_sequence
+    ),
+  )
+end
+
+"""Perform an SVD on p*q*o (with identity environments) to get an initial guess for the full update"""
+function initial_guess_pprime_qprime(p::ITensor, q::ITensor, o::ITensor; maxdim=nothing)
+  p_out, q_out = factorize(noprime(p * q * o), inds(p); maxdim)
+
+  cur_ind = commonind(p_out, q_out)
+  replaceind!(p_out, cur_ind, replacetags(cur_ind, tags(cur_ind), tags(commonind(p, q))))
+  replaceind!(q_out, cur_ind, replacetags(cur_ind, tags(cur_ind), tags(commonind(p, q))))
+
+  return p_out, q_out
+end
+
+function M_p(
+  envs::Vector{ITensor}, p_q_tensor::ITensor, apply_tensor::ITensor; opt_sequence=nothing
+)
+  return noprime(
+    ITensors.contract(
+      vcat(
+        ITensor[p_q_tensor, noprime!(prime(dag(p_q_tensor)); tags="Site"), apply_tensor],
+        envs,
+      );
+      sequence=opt_sequence,
+    ),
+  )
+end
+
+"""Calculate the overlap of the gate acting on the previous p and q versus the new p and q in the presence of environments. This is the cost function to minimise"""
+function fidelity(
+  envs::Vector{ITensor},
+  p_cur::ITensor,
+  q_cur::ITensor,
+  p_prev::ITensor,
+  q_prev::ITensor,
+  gate::ITensor,
+)
+  term1_tns = vcat(
+    [
+      p_prev,
+      q_prev,
+      prime(dag(p_prev)),
+      prime(dag(q_prev)),
+      swapprime(gate * swapprime(dag(gate), 0, 2), 2, 1),
+    ],
+    envs,
+  )
+  term1 = ITensors.contract(
+    term1_tns; sequence=ITensors.optimal_contraction_sequence(term1_tns)
+  )
+  term2_tns = vcat(
+    [
+      p_cur,
+      q_cur,
+      noprime(prime(dag(p_cur)); tags="Site"),
+      noprime(prime(dag(q_cur)); tags="Site"),
+    ],
+    envs,
+  )
```

Comment:
This kind of code using too much priming and tag filtering raises red flags for me. Is there a way to do this by analyzing the graph/environment and using graph and index set operations?

Reply:
I will clean it up and try to do it more naturally.